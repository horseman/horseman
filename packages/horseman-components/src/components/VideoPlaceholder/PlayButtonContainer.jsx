import styled, { css } from "styled-components";
import PropTypes from "prop-types";
import media from "../../styleUtils/media";

/**
 * Container for the play button that will position it centered left or right.
 * This container is absolutely positioned within the VideoPlaceholder.
 *
 * NOTE: in mobile the play button is centered by default; left & right are only for bigger screens
 */
const PlayButtonContainer = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);

  ${media.mdAndUp(css`
    ${props => {
      switch (props.position) {
        case "left": {
          return css`
            left: 0;
            transform: translate(-50%, -50%);
          `;
        }
        case "right": {
          return css`
            right: 0;
            transform: translate(50%, -50%);
          `;
        }
        default: {
          return css`
            left: 50%;
            transform: translate(-50%, -50%);
          `;
        }
      }
    }};
  `)};
`;

PlayButtonContainer.defaultProps = {
  position: "center",
};

PlayButtonContainer.propTypes = {
  /**
   * Where should the play button be positioned within this container
   */
  position: PropTypes.oneOf(["left", "center", "right"]),
};

export default PlayButtonContainer;
