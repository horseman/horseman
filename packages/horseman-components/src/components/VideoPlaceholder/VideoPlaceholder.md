This is a documentation Stub. Each component is required to be documented with
examples

### Examples
Example of a default VideoPlaceholder. Update the `playPosition` prop to see
how that is able to shift 'left', 'right', or 'center'.
```
const imageObject = require('styleguide/fixtures/imageObject').default;

<VideoPlaceholder
  playPosition="center"
  bgImage={imageObject}
  video={{
    id: '9ytei6bu7kQ',
    source: 'youtube',
  }}
/>
```
Example of a VideoPlaceholder filling it's relatively positioned container.


```
const imageObject = require('styleguide/fixtures/imageObject').default;

<div style={{position: 'relative', overflow: 'visible', height: '400px', maxWidth: '300px' }}>
  <VideoPlaceholder
    playPosition="right"
    bgImage={imageObject}
    video={{
      id: '9ytei6bu7kQ',
      source: 'youtube',
    }}
    fill
  />
</div>
```
