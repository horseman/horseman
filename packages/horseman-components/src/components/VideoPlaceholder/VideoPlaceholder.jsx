import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import { imageType } from "../../types";

import ResponsiveBackgroundImage from "../ResponsiveBackgroundImage";
import PlayButtonContainer from "./PlayButtonContainer";

const StyledVideoPlaceholder = styled(({ fill, ...rest }) => (
  <ResponsiveBackgroundImage {...rest} />
))`
  /**
   * establish a 16/9 aspect ratio for the default video, if we otherwise don't
   * want to fill the container.
   */
  padding-top: ${({ fill }) => (fill ? 0 : 9 / 16 * 100)}%;
  overflow: visible;
  height: 100%;
  position: relative;
`;

const VideoPlaceholder = ({ bgImage, playPosition, playButton, fill }) => (
  <StyledVideoPlaceholder {...bgImage} fill={fill}>
    <PlayButtonContainer position={playPosition}>
      {playButton}
    </PlayButtonContainer>
  </StyledVideoPlaceholder>
);

VideoPlaceholder.defaultProps = {
  fill: false,
  playPosition: "center",
};

VideoPlaceholder.propTypes = {
  /**
   * Pass true to make the video placeholder fill it's container and not
   * conform to a 16/9 aspect ratio
   */
  fill: PropTypes.bool,

  /**
   * Where should the play button be positioned within this placeholder.
   */
  playPosition: PropTypes.oneOf(["left", "right", "center"]),

  /**
   * The button to be rendered representing "play".
   */
  playButton: PropTypes.node.isRequired,

  /**
   * What image will be used for the video placeholder image
   */
  bgImage: PropTypes.shape({ ...imageType }).isRequired,
};

export default VideoPlaceholder;
