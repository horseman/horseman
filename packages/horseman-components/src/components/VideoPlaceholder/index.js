import VideoPlaceholder from "./VideoPlaceholder";

export default VideoPlaceholder;
