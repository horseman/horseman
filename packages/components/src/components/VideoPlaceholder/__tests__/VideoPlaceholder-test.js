import React from "react";
import { shallow } from "enzyme";

import VideoPlaceholder from "../";

describe("VideoPlaceholder", () => {
  const image = {
    src: "default",
    srcset: [
      {
        src: "default",
        width: 100,
      },
      {
        src: "larger",
        width: 200,
      },
      {
        src: "largest",
        width: 300,
      },
    ],
  };
  test("renders correctly", () => {
    const wrapper = shallow(
      <VideoPlaceholder playButton={<div />} bgImage={image} />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  test("can fill it's container", () => {
    const wrapper = shallow(
      <VideoPlaceholder playButton={<div />} bgImage={image} fill />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});
