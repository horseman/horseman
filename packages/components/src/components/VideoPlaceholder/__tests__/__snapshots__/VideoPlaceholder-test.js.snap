// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlaceholder can fill it's container 1`] = `
.c0 {
  padding-top: 0%;
  overflow: visible;
  height: 100%;
  position: relative;
}

<Component
  className="c0"
  fill={true}
  src="default"
  srcset={
    Array [
      Object {
        "src": "default",
        "width": 100,
      },
      Object {
        "src": "larger",
        "width": 200,
      },
      Object {
        "src": "largest",
        "width": 300,
      },
    ]
  }
>
  <styled.div
    position="center"
  >
    <div />
  </styled.div>
</Component>
`;

exports[`VideoPlaceholder renders correctly 1`] = `
.c0 {
  padding-top: 56.25%;
  overflow: visible;
  height: 100%;
  position: relative;
}

<Component
  className="c0"
  fill={false}
  src="default"
  srcset={
    Array [
      Object {
        "src": "default",
        "width": 100,
      },
      Object {
        "src": "larger",
        "width": 200,
      },
      Object {
        "src": "largest",
        "width": 300,
      },
    ]
  }
>
  <styled.div
    position="center"
  >
    <div />
  </styled.div>
</Component>
`;
